File: July 2024/09-07-2024/Closest Three Sum.py
#User function Template for python3

# arr    : list of integers denoting the elements of the array
# target : as specified in the problem statement

class Solution:
    def threeSumClosest (self, arr, target):
    # Your Code Start
            # Sort the array
        arr.sort()
        
        n = len(arr)
        close_sum = arr[0] + arr[1] + arr[2]  # Initialize close_sum with the first triplet
        
        # Iterate through the array
        for i in range(n - 2):
            l = i + 1
            r = n - 1
            
            # Use two-pointer technique
            while l < r:
                curr_sum = arr[i] + arr[l] + arr[r]
                
                # If the current sum is exactly equal to the target, return it
                if curr_sum == target:
                    return curr_sum
                
                # Update the close_sum if the current sum is closer to the target
                if abs(curr_sum - target) < abs(close_sum - target):
                    close_sum = curr_sum
                elif abs(curr_sum - target) == abs(close_sum - target) and curr_sum > close_sum:
                    # Prefer the maximum sum if the difference is the same
                    close_sum = curr_sum
                
                # Move the pointers
                if curr_sum < target:
                    l += 1
                else:
                    r -= 1
        
        return close_sum
    #Your Code End


#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__':
    t = int(input())

    for _ in range(t):
        A = list(map(int, input().strip().split()))
        k = int(input())
        ob = Solution()
        print(ob.threeSumClosest(A, k))

# } Driver Code Ends

File: July 2024/09-07-2024/Closest Three Sum.java
//{ Driver Code Starts
// Initial Template for Java

import java.io.*;
import java.util.*;

public class GFG {
    public static void main(String args[]) throws IOException {
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
        PrintWriter out = new PrintWriter(System.out);

        int t = Integer.parseInt(in.readLine().trim());
        while (t-- > 0) {
            String line = in.readLine();
            String[] tokens = line.split(" ");

            // Create an ArrayList to store the integers
            ArrayList<Integer> array = new ArrayList<>();

            // Parse the tokens into integers and add to the array
            for (String token : tokens) {
                array.add(Integer.parseInt(token));
            }

            int[] arr = new int[array.size()];
            int idx = 0;
            for (int i : array) arr[idx++] = i;

            int key = Integer.parseInt(in.readLine().trim());

            out.println(new Solution().threeSumClosest(arr, key));
        }
        out.close();
    }
}
// } Driver Code Ends


// User function Template for Java

class Solution {
    static int threeSumClosest(int[] array, int target) {
        //Your Code Start
       // Sort the array
        Arrays.sort(array);
        
        int n = array.length;
        int closeSum = array[0] + array[1] + array[2]; // Initialize closeSum with the first triplet
        
        // Iterate through the array
        for (int i = 0; i < n - 2; ++i) {
            int l = i + 1;
            int r = n - 1;
            
            // Use two-pointer technique
            while (l < r) {
                int currSum = array[i] + array[l] + array[r];
                
                // If the current sum is exactly equal to the target, return it
                if (currSum == target) {
                    return currSum;
                }
                
                // Update the closeSum if the current sum is closer to the target
                if (Math.abs(currSum - target) < Math.abs(closeSum - target)) {
                    closeSum = currSum;
                } else if (Math.abs(currSum - target) == Math.abs(closeSum - target) && currSum > closeSum) {
                    // Prefer the maximum sum if the difference is the same
                    closeSum = currSum;
                }
                
                // Move the pointers
                if (currSum < target) {
                    ++l;
                } else {
                    --r;
                }
            }
        }
        
        return closeSum;
        //Your Code End
    
    }
}
